components/ee/ws-scheduler/pkg/scaler/controller.go

```diff
@@ -0,0 +1,284 @@
+				diff := f.Target - c.Ghost
+				res <- diff
+			}
+		}
+	}()
+	return res
+}
+
+// TimeOfDay is a time during the day. It unmarshals from JSON as hh:mm:ss string.
+type TimeOfDay time.Time
+
+// UnmarshalJSON unmarshales a time of day
+func (t *TimeOfDay) UnmarshalJSON(data []byte) error {
+	res, err := time.Parse("15:04:05", string(data))
+	if err != nil {
+		return err
+	}
+	*t = TimeOfDay(res)
+	return nil
+}
+
+// SwitchedSetpoint is a setpoint valid from a particular time in the day
+type SwitchedSetpoint struct {
+	Time     TimeOfDay `json:"time"`
+	Setpoint int       `json:"setpoint"`
+}
+
+// NewSwitchedSetpointController creates a new SwitchedSetpointController
+func NewSwitchedSetpointController(defaultSetpoint int, setpoints []SwitchedSetpoint) (*SwitchedSetpointController, error) {
+	if defaultSetpoint < 0 {
+		return nil, xerrors.Errorf("defaultSetpoint must be >= 0")
+	}
+
+	sort.Slice(setpoints, func(i, j int) bool { return time.Time(setpoints[i].Time).Before(time.Time(setpoints[j].Time)) })
+	return &SwitchedSetpointController{
+		DefaultSetpoint: defaultSetpoint,
+		Setpoints:       setpoints,
+		NewTicker:       newDefaultTicker(1 * time.Minute),
+		Now:             time.Now,
+		SetpointChanged: func(old, new int) {
+			if old == new {
+				return
+			}
+
+			log.WithField("new", new).WithField("old", new).Info("timed function controller target change")
+		},
+	}, nil
+}
+
+// SwitchedSetpointController is like the ConstantSetpointController but with different
+// setpoints throughout the day.
+type SwitchedSetpointController struct {
+	DefaultSetpoint int
+	Setpoints       []SwitchedSetpoint
+
+	NewTicker       func() (c <-chan time.Time, stop func())
+	SetpointChanged func(old, new int)
+	Now             func() time.Time
+}
+
+// Control starts this controller
+func (c *SwitchedSetpointController) Control(ctx context.Context, workspaceCount <-chan WorkspaceCount) (ghostDelta <-chan int) {
+	res := make(chan int)
+
+	setpoint := c.DefaultSetpoint
+	if csp := c.findSwitchpoint(c.Now()); csp != nil {
+		c.SetpointChanged(setpoint, csp.Setpoint)
+		setpoint = csp.Setpoint
+	}
+
+	tick, stop := c.NewTicker()
+	go func() {
+		defer stop()
+		defer close(res)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case t := <-tick:
+				nsp := -1
+				if csp := c.findSwitchpoint(t); csp != nil {
+					nsp = csp.Setpoint
+				} else {
+					nsp = c.DefaultSetpoint
+				}
+				c.SetpointChanged(setpoint, nsp)
+				setpoint = nsp
+			case c := <-workspaceCount:
+				diff := setpoint - c.Ghost
+				res <- diff
+			}
+		}
+	}()
+	return res
+}
+
+func (c *SwitchedSetpointController) findSwitchpoint(t time.Time) *SwitchedSetpoint {
+	if len(c.Setpoints) == 0 {
+		return nil
+	}
+
+	tod := time.Date(0, 1, 1, t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), t.Location())
+	for i, sp := range c.Setpoints {
+		spt := time.Time(sp.Time)
+		if tod.Equal(spt) {
+			return &sp
+		}
+		if tod.After(spt) {
+			continue
+		}
+		if i == 0 {
```

Comment:
And what does this mean? It always make sure we never get passed the first setpoint in the array?
Maybe I'm not seeing the logic here, in that case: please add a comment! :-D

Reply:
see above :)

Reply:
Now it makes sense. :) I'd prefer
```
if tod.Before(time.Time(c.Setpoints[0].Time)) {
  return nil
}
```
which would make the loop "cleaner" (less special cases). But that's a matter of taste.

Reply:
indeed ... I did have this in a version, but replaced it with the `i == 0` case because it explicitly guards against the `i = 0; c.Setpoints[i - 1]` case. With the the `tod.Before` clause that connection isn't as obvious.

All in all I'm not terribly happy with this function. It's too hard to understand for what it does.